src/Columns/ColumnSparse.cpp

```diff
@@ -0,0 +1,762 @@
+    auto & res_offsets_data = assert_cast<ColumnUInt64 &>(*res_offsets).getData();
+    auto res_values = values->cloneEmpty();
+    res_values->insertDefault();
+
+    auto offset_it = begin();
+    for (size_t i = 0; i < _size; ++i, ++offset_it)
+    {
+        if (!offset_it.isDefault())
+        {
+            size_t replicate_size = replicate_offsets[i] - replicate_offsets[i - 1];
+            res_offsets_data.reserve(res_offsets_data.size() + replicate_size);
+            for (size_t row = replicate_offsets[i - 1]; row < replicate_offsets[i]; ++row)
+            {
+                res_offsets_data.push_back(row);
+                res_values->insertFrom(*values, offset_it.getValueIndex());
+            }
+        }
+    }
+
+    return ColumnSparse::create(std::move(res_values), std::move(res_offsets), replicate_offsets.back());
+}
+
+void ColumnSparse::updateHashWithValue(size_t n, SipHash & hash) const
+{
+    values->updateHashWithValue(getValueIndex(n), hash);
+}
+
+void ColumnSparse::updateWeakHash32(WeakHash32 & hash) const
+{
+    if (hash.getData().size() != _size)
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Size of WeakHash32 does not match size of column: "
+        "column size is {}, hash size is {}", _size, hash.getData().size());
+
+    auto offset_it = begin();
+    auto & hash_data = hash.getData();
+    for (size_t i = 0; i < _size; ++i, ++offset_it)
+    {
+        size_t value_index = offset_it.getValueIndex();
+        auto data_ref = values->getDataAt(value_index);
+        hash_data[i] = ::updateWeakHash32(reinterpret_cast<const UInt8 *>(data_ref.data), data_ref.size, hash_data[i]);
+    }
+}
+
+void ColumnSparse::updateHashFast(SipHash & hash) const
+{
+    values->updateHashFast(hash);
+    offsets->updateHashFast(hash);
+    hash.update(_size);
+}
+
+void ColumnSparse::getExtremes(Field & min, Field & max) const
+{
+    if (_size == 0)
+    {
+        values->get(0, min);
+        values->get(0, max);
+        return;
+    }
+
+    if (getNumberOfDefaults() == 0)
+    {
+        size_t min_idx = 1;
+        size_t max_idx = 1;
+
+        for (size_t i = 2; i < values->size(); ++i)
+        {
+            if (values->compareAt(i, min_idx, *values, 1) < 0)
+                min_idx = i;
+            else if (values->compareAt(i, max_idx, *values, 1) > 0)
+                max_idx = i;
+        }
+
+        values->get(min_idx, min);
+        values->get(max_idx, max);
+        return;
+    }
+
+    values->getExtremes(min, max);
+}
+
+void ColumnSparse::getIndicesOfNonDefaultRows(IColumn::Offsets & indices, size_t from, size_t limit) const
+{
+    const auto & offsets_data = getOffsetsData();
+    const auto * start = from ? std::lower_bound(offsets_data.begin(), offsets_data.end(), from) : offsets_data.begin();
+    const auto * end = limit ? std::lower_bound(offsets_data.begin(), offsets_data.end(), from + limit) : offsets_data.end();
+
+    indices.insert(start, end);
```

Comment:
Either we accept that `getIndicesOfNonDefaultRows` may return not all indexes (need comment) or we need to check values (or insertRange/insertRangeFrom for full column is implemented not correctly).

Reply:
I've correct the implemetation. Now defaults can't be stored in `values` column.